argon2_out_of_process_support.go

```diff
@@ -0,0 +1,990 @@
+			case <-tmb.Dying():
+				// The tomb began dying before we finished sending the request (reqChan is blocking).
+				return tomb.ErrDying
+			}
+
+			// Wait for the response from the remote process.
+			select {
+			case <-time.NewTimer(timeout).C: // Give it up to the time defined by the timeout
+				return errors.New("timeout waiting for watchdog response from remote process")
+			case rsp := <-rspChan:
+				// We got a response from the remote process.
+				if err := rsp.Err(); err != nil {
+					// We got an error response, so just return the error.
+					return rsp.Err()
+				}
+				if !bytes.Equal(rsp.WatchdogResponse, expectedWatchdogResponse) {
+					// We got an unexpected response, so return an error.
+					return errors.New("unexpected watchdog response value from remote process")
+				}
+				// The response was good so save the value for the next iteration.
+				lastWatchdogResponse = rsp.WatchdogResponse
+			case <-tmb.Dying():
+				// The loop won't run again
+			}
+		}
+		return tomb.ErrDying
+	}
+}
+
+// NoArgon2OutOfProcessWatchdogMonitor is an implmenentation of Argon2OutOfProcessWatchdogMonitor that
+// provides no watchdog functionality. It is paired with [NoArgon2OutOfProcessWatchdogHandler] on the
+// remote side. It holds the watchdog goroutine in a parked state for the lifetime of the tomb.
+func NoArgon2OutOfProcessWatchdogMonitor() Argon2OutOfProcessWatchdogMonitor {
+	return func(tmb *tomb.Tomb, reqChan chan<- *Argon2OutOfProcessRequest, rspChan <-chan *Argon2OutOfProcessResponse) error {
+		select {
+		case <-tmb.Dying():
+			return tomb.ErrDying
+		case <-rspChan:
+			return errors.New("unexpected watchdog response")
+		}
+	}
+}
+
+type outOfProcessArgon2KDFImpl struct {
+	newHandlerCmd func() (*exec.Cmd, error)
+	timeout       time.Duration
+	watchdog      Argon2OutOfProcessWatchdogMonitor
+}
+
+func (k *outOfProcessArgon2KDFImpl) sendRequestAndWaitForResponse(req *Argon2OutOfProcessRequest) (*Argon2OutOfProcessResponse, error) {
+	// Use ther user-supplied function to create a new *exec.Cmd structure.
+	cmd, err := k.newHandlerCmd()
+	if err != nil {
+		return nil, fmt.Errorf("cannot create new command: %w", err)
+	}
+
+	// Configure an OS pipe for stdin for sending requests.
+	stdinPipe, err := cmd.StdinPipe()
+	if err != nil {
+		// This doesn't fail once the OS pipe is created, so there's no
+		// cleanup to do on failure paths.
+		return nil, fmt.Errorf("cannot create stdin pipe: %w", err)
+	}
+
+	// Configure an OS pipe for stdout for receiving responses.
+	stdoutPipe, err := cmd.StdoutPipe()
+	if err != nil {
+		// This doesn't fail once the OS pipe is created, so there's no
+		// cleanup to do on failure paths other than closing the stdinPipe
+		stdinPipe.Close()
+		return nil, fmt.Errorf("cannot create stdout pipe: %w", err)
+	}
+
+	// Start the remote process.
+	if err := cmd.Start(); err != nil {
+		return nil, fmt.Errorf("cannot start handler process: %w", err)
```

Comment:
do the pipes need to be closed?

Reply:
They do. Both of them get closed at the end of the function with comments explaining that it's required to unblock the remote side so that it can shut down cleanly.

Reply:
What I mean is that the pipes were created for stdin/stdout by now. However, the code is exiting early here and there was no defer closing them added before. Actually, the Go documentation isn't exactly clear on what happens. Looking at source code for 1.23.4 the pipes should be closed automatically if Start() failed? https://cs.opensource.google/go/go/+/refs/tags/go1.23.4:src/os/exec/exec.go;l=631-635 makes me wonder if this behavior is consistent with earlier releases.

Reply:
Ooh, good catch - yes, I'm not sure what happens in the case where `Start` fails. I'll take a look at go 1.18 and see if the behaviour is the same.

Reply:
Having a look at go 1.18 and the latest stable, it looks like both ends of each pipes we created get closed by `exec.Cmd.Start` if it fails, which is good. Where there was a bug in this code though is that if we get an error back from `exec.Cmd.StdoutPipe` after already calling `exec.Cmd.StdinPipe`, we closed only the parent side of the pipe. I've fixed this to make sure that both ends get closed now.